kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1", "output-encoding":"json"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. Once an analysis has started, it is designated as **pending**
+    until the analyzer sends a `done` notification (see below). The driver MUST
+    report ProtocolError and close the input channel if the analyzer issues an
+    `analyze` request while another analysis is pending.
+    
+    The reply includes the input file(s) on which to begin analysis, as well as
+    the output file to which to emit analysis output. It reflects some of the
+    kythe.proto.CompilationUnit structure. The driver (and driver environment)
+    MUST make contents available for each file referenced in the compilation
+    record at the file system path specified in the compilation record.
+
+    The output file should contain the content appropriate to the
+    analysis-type, encoded as specified in the `InitRequest`. For Kythe index
+    analyses, this would be `kythe.common.Entry` objects. Other analyzers might
+    emit a "findings" structure.
+
+    Note that contents need not be literally copied. An overlay filesystem,
+    hardlinks, symlinks, or other may be used.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":2,"method":"analyze","params":{"types":["kythe/indexer/go"]}}
+        « {"jsonrpc":"2.0","id":2","result":{"input":"path/to/main.go", "output":"path/to/output.json"}}
+
+-   `vname(VNameRequest)`
+
+    Requests the complete VName to emit for the specified path and signature.
+
+-   `log(LogRequest)`
+
+    A notification from the analyzer to the environment, containing a message
+    string to be logged in association with the current analysis. The
+    environment does not reply.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"log","params":{"message":{...}}}
+
+-   `done(LogRequest)`
+
+    A notification from the analyzer to the environment, signifying that the
+    analyzer has completed work on the specified analysis ID. The environment
+    does not reply. After receiving this notification, the driver MUST discard
+    any further requests for the analysis ID (and return an error, if
+    appropriate) unless and until it issues a new analysis reusing that ID.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"done","params":{"message":"success"}}
+
+### Errors
```

Comment:
How does the analyzer communicate that it had a unrecoverable internal error?

Reply:
The current proposal is that it terminates with a non-zero exit code. Do you see cases where an analyzer would want to abort the current analysis (e.g. due to unrecoverable malformed input) but continue to be able to request analyses, without restarting? Is that common enough that it warrants additional complexity?

Reply:
I more had in mind that the analyzer could return a stack trace or other rich description of the error condition, which could be cleaner than scraping stderr.

Reply:
While some analyzers may be able to return structured data for internal errors quite a few of them will segfault and (if lucky) print a stack trace on stderr, so if we want to capture that we'll have to scrape stderr regardless.

Reply:
There are different ways an analyzer can fail. If it fails under control, it can report an error in-channel, which along with the `"done"` message. If it fails out-of-control, however, the program will simply exit. The protocol needs to account for both.

Even when an analyzer _tries_ to control failures, it can't always. Even in a managed runtime the OS may kill a process for exceeding resource limits or poor use of FFI. Moreover some compiler toolchains do not manage memory well (GHC, Python, Dart, and SBCL for some examples we've encountered in the past) and will pile up state and garbage if you re-use the analyzer without an exit in between.

So at some point you have to deal with the fact that the program either exits uncontrollably, or exits because it is simpler to exit and restart than to rewrite the compiler. I think the protocol should deal with that as simply as possible—at which point reading the exit code is probably the best we can hope for. Scraping stderr is probably not useful for an in-band message, though it makes sense to capture stderr for logging purposes.